pkg/liqo-controller-manager/namespace-controller/namespace_controller.go

```diff
@@ -0,0 +1,270 @@
+				return ctrl.Result{}, err
+			}
+
+			for k := range removeMappings {
+				delete(removeMappings, k)
+			}
+
+		} else {
+
+			klog.Info(" Watch for virtual-nodes labels")
+
+			// 2.b Iterate on all virtual nodes labels, if the namespace has all the requested labels, is necessary to
+			// offload it onto remote cluster associated with the virtual node
+			nodes := &corev1.NodeList{}
+			if err := r.List(context.TODO(), nodes, client.MatchingLabels{liqo_controller_manager.TypeLabel: liqo_controller_manager.TypeNode}); err != nil {
+				klog.Error(err, " --> Unable to List all virtual nodes")
+				return ctrl.Result{}, client.IgnoreNotFound(err)
+			}
+
+			if len(nodes.Items) == 0 {
+				klog.Info(" No VirtualNode at the moment")
+				return ctrl.Result{}, nil
+			}
+
+			for _, node := range nodes.Items {
+				i := 0
+				dim := len(node.Labels)
+				id := node.Annotations[liqo_controller_manager.VirtualNodeClusterId]
+				for key := range node.GetLabels() {
+					i++
+					if strings.HasPrefix(key, offloadingPrefixLabel) {
+						if _, ok = namespace.GetLabels()[key]; !ok {
+							klog.Infof(" Namespace '%s' cannot be offloaded on remote cluster: %s", namespace.GetName(), id)
+							break
+						}
+					}
+
+					if i == dim {
+						klog.Infof(" Offload namespace '%s' on remote cluster: %s", namespace.GetName(), id)
+						if err := r.createRemoteNamespace(namespace, remoteNamespaceName, removeMappings[id]); err != nil {
+							return ctrl.Result{}, err
+						}
+						delete(removeMappings, id)
+					}
+
+				}
+			}
+		}
+
+	}
+
+	if len(removeMappings) > 0 {
+		klog.Info(" Delete all unnecessary mapping in NamespaceMaps")
+		if err := r.removeRemoteNamespaces(namespace.GetName(), removeMappings); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func mappingLabelUpdate(oldLabels map[string]string, newLabels map[string]string) bool {
+	ret := false
+	if val1, ok := oldLabels[mappingLabel]; ok {
+		ret = val1 != newLabels[mappingLabel]
+	}
+	return ret
+}
+
+func mappingLabelPresence(labels map[string]string) bool {
+	_, ok := labels[mappingLabel]
+	return ok
+}
+
+// Events not filtered:
+// 1 -- deletion timestamp is updated on a relevant namespace (only that ones with my finalizer)
+// 2 -- add/delete labels, and mappingLabel is present before or after the update
+// 3 -- update the value of mappingLabel label only
+// 4 -- add namespace with at least mappingLabel
+func manageLabelPredicate() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if !(e.MetaNew.GetDeletionTimestamp().IsZero()) && slice.ContainsString(e.MetaNew.GetFinalizers(), namespaceFinalizer, nil) {
+				return true
+			}
+			oldLabels := e.MetaOld.GetLabels()
+			newLabels := e.MetaNew.GetLabels()
+			return ((len(oldLabels) != len(newLabels)) && (mappingLabelPresence(oldLabels) || mappingLabelPresence(newLabels))) || mappingLabelUpdate(oldLabels, newLabels)
```

Comment:
Feel free to use multi-line. Besides, I suggest avoiding assigning labels to brand-new variables (practically  useless)

Reply:
Thank you